### PR TITLE
AStyleLintBear.py: Add a linter bear for AStyle

### PR DESCRIPTION
**Pull request by ivange94**
Add a linter bear for the Artistic Style linter

Closes https://github.com/coala/coala-bears/issues/388

### CODE REVIEW

File: bears/c_languages/AStyleLintBear.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    For more information, consult <http://astyle.sourceforge.net/astyle.html>.
+    """
+
+    LANGUAGES = {'C, C++, Objective-C, C#, Java'}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    CAN_DETECT = {'Formatting'}
+
```
Comment (vijeth-aradhya) [inline bears/c_languages/AStyleLintBear.py]: Add `REQUIREMENTS` here
And also `ASCIINEMA_URL`

File: bears/c_languages/AStyleLintBear.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        """
+        :param astyle_style:
+             Defines the bracket style to use. If no bracket style is
+             specified, the default 'allman' is used.
+        :param astyle_indent:
+             The indentation type to use. It can either be spaces or tabs.
+        :param astyle_astylerc:
+             Optional configuration file for astyle to read configs.
+        :param astyle_indent_size:
+            Number of spaces per indentation level.
```
Comment (nemaniarjun) [inline bears/c_languages/AStyleLintBear.py]: Indentation error

File: tests/c_languages/AStyleLintBearTest.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
+from bears.c_languages.AStyleLintBear import AStyleLintBear
+from coalib.testing.LocalBearTestHelper import verify_local_bear
+import os
```
Comment (nemaniarjun) [inline tests/c_languages/AStyleLintBearTest.py]: Google  PEP8 imports.

File: tests/c_languages/AStyleLintBearTest.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+test_file4 = """
+int main()
+{
+        int a=23;
+        return 0;
+}"""
+
+
+def get_testfile_path(name):
+    return os.path.join(os.path.dirname(__file__), 'astyle_test_files',  name)
```
Comment (nemaniarjun) [inline tests/c_languages/AStyleLintBearTest.py]: No need for a function

File: bears/c_languages/AStyleLintBear.py (hunk 1 above)
Comment (nemaniarjun) [inline bears/c_languages/AStyleLintBear.py]: Also doesn't this correct the file? If so, use can fix also.

**Comment by nemaniarjun** [comment]
Since it also supports Java, we might not want to keep it in C_languages. Also change your commit shortlog to `C_languages: Add AstyleLintBear`

File: bears/c_languages/AStyleLintBear.py (hunk 2 above)
Comment (ivange94) [inline bears/c_languages/AStyleLintBear.py]: I thought running 'coala' would fix something like that.

File: tests/c_languages/AStyleLintBearTest.py (hunk 4 above)
Comment (ivange94) [inline tests/c_languages/AStyleLintBearTest.py]: Is there a problem with having that function?

File: bears/c_languages/AStyleLintBear.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+from coalib.bearlib.abstractions.Linter import linter
+from coalib.bearlib.spacing.SpacingHelper import SpacingHelper
+from dependency_management.requirements.DistributionRequirement import (
+    DistributionRequirement)
```
Comment (ivange94) [inline bears/c_languages/AStyleLintBear.py]: Not sure what the problem here is.

File: bears/c_languages/AStyleLintBear.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        :param astyle_indent_size:
+            Number of spaces per indentation level.
+        """
+        args = ('--dry-run', '--suffix=none')
+        if astyle_style:
+            args += ('--style=' + astyle_style,)
+        if astyle_indent:
+            indent = '--indent=' + astyle_indent
+            indent += '=' + \
+                str(astyle_indent_size) if astyle_indent_size else ''
```
Comment (Makman2) [inline bears/c_languages/AStyleLintBear.py]: please use braces to wrap your expression:
```
indent += ('=' +
    str(astyle.....)... )
```

Comment (Makman2) [inline bears/c_languages/AStyleLintBear.py]: btw if you have no indent size given, is `--indent=style=` a valid option?

File: bears/c_languages/AStyleLintBear.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            args += (indent,)
+        if astyle_astylerc:
+            args += ('--options=' + astyle_astylerc,)
```
Comment (Makman2) [inline bears/c_languages/AStyleLintBear.py]: actually you can use a list instead of a tuple for `args`, linter doesn't alter the args returned + you don't need them anyway after that call. Makes stuff slightly faster, as appending to tuples copies the old tuple every time^^

File: tests/c_languages/AStyleLintBearTest.py (hunk 4 above)
Comment (Makman2) [inline tests/c_languages/AStyleLintBearTest.py]: it's used, so it has definitely a reason^^

File: bears/c_languages/AStyleLintBear.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+class AStyleLintBear:
+    """
+    Artistic Style is a source code indenter, formatter,
+    and beautifier for the C, C++, C++/CLI, Objective‑C,
+    C# and Java programming languages.
+
+    For more information, consult <http://astyle.sourceforge.net/astyle.html>.
+    """
+
+    LANGUAGES = {'C, C++, Objective-C, C#, Java'}
```
Comment (Makman2) [inline bears/c_languages/AStyleLintBear.py]: naah that's not how sets work :D -->
```
LANGUAGES = {'C', 'C++', 'Objective-C', 'C#', 'Java'}
```

**Comment by nemaniarjun** [comment]
Astlye can be installed using apt-get. Confirmed. Downloads the version 2.05, latest is 2.06.
Lets use apt for travis.

**Comment by jayvdb** [comment]
@ivange94 , could you please rebase.  I suspect the main errors in your builds are caused by a problem which has been fixed.

File: bears/c_languages/AStyleLintBear.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    For more information, consult <http://astyle.sourceforge.net/astyle.html>.
+    """
+
+    LANGUAGES = {'C', 'C++', 'Objective-C', 'C#', 'Java'}
+    REQUIREMENTS = {DistributionRequirement(apt_get='astyle')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    CAN_DETECT = {'Formatting'}
+    CAN_FIX = {'Formatting'}
```
Comment (Asnelchristian) [inline bears/c_languages/AStyleLintBear.py]: CAN_FIX is enough

Comment (Asnelchristian) [inline bears/c_languages/AStyleLintBear.py]: what i mean is, there is no need for ``CAN_DETECT``

File: bears/c_languages/AStyleLintBear.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         astyle_style: str='',
+                         astyle_indent: str='',
+                         astyle_astylerc: str='',
+                         astyle_indent_size: int=0):
```
Comment (Asnelchristian) [inline bears/c_languages/AStyleLintBear.py]: use ``indent_style`` for consistency

File: bears/c_languages/AStyleLintBear.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    CAN_DETECT = {'Formatting'}
+    CAN_FIX = {'Formatting'}
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         astyle_style: str='',
+                         astyle_indent: str='',
+                         astyle_astylerc: str='',
```
Comment (Asnelchristian) [inline bears/c_languages/AStyleLintBear.py]: maybe ``astyle_config`` for consistency, like ``yamllint_config``, ``tslint_config``, ``scalalint_config`` etc....

File: bears/c_languages/AStyleLintBear.py (hunk 10 above)
Comment (Asnelchristian) [inline bears/c_languages/AStyleLintBear.py]: the same goes for the other settings, you should rename them with consistency in mind

File: bears/c_languages/AStyleLintBear.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    CAN_DETECT = {'Formatting'}
+    CAN_FIX = {'Formatting'}
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         astyle_style: str='',
+                         astyle_indent: str='',
```
Comment (Asnelchristian) [inline bears/c_languages/AStyleLintBear.py]: maybe ``use_spaces: bool=None``

File: bears/c_languages/AStyleLintBear.py (hunk 10 above)
Comment (ivange94) [inline bears/c_languages/AStyleLintBear.py]: All settings had consistency in mind. The question is consistency with what. I designed them to be consistent with the underlying tool they are wrapping(i.e astyle) so users of these tools will not have to spend time learning new names. But you suggestions mean consistency with other coala bears. Is that preferable?

Comment (Makman2) [inline bears/c_languages/AStyleLintBear.py]: consistency with our other bears is preferred :+1: :) Then you can easily re-use settings from a coafile :+1:

**Comment by Asnelchristian** [comment]
wrt to the other settings,  settings of the other bears e.g ``use_spaces``,
several bears use this setting ( which simply require spaces to be used for
indentation ) so in you case you might wanna use the same name for
``astyle_indent``

2017-02-01 11:44 GMT+01:00 Ivange Larry <notifications@github.com>:

File: .travis.yml [hunk 13]
```diff
@@ -93,6 +94,9 @@ before_install:
       sed -e '/^coala/d' requirements.txt > requirements.new
       mv requirements.new requirements.txt
     fi
+  # Install astyle linter
+  - sudo apt-get -qq update
+  - sudo apt-get install -y astyle
```
Comment (jayvdb) [inline .travis.yml]: why is this necessary?

File: bears/c_languages/AStyleLintBear.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    CAN_FIX = {'Formatting'}
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         astyle_style: str='',
+                         use_spaces: bool=None,
+                         astyle_config: str='',
+                         indent_size: int=0):
+        """
+        :param astyle_style:
```
Comment (jayvdb) [inline bears/c_languages/AStyleLintBear.py]: does it only apply to brackets?  if so `astyle_bracket_style` would be better ?

File: bears/c_languages/AStyleLintBear.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+            Defines the bracket style to use. If no bracket style is
+            specified, the default 'allman' is used.
+        :param use_spaces:
+            If None, no indentation is specified and astyle uses a default.
+            If True, astyle uses spaces or if explicitly set to False, astyle
+            uses tabs for indentation.
+        :param astyle_config:
+            Optional configuration file for astyle to read configs.
```
Comment (jayvdb) [inline bears/c_languages/AStyleLintBear.py]: I think this can be achieved using `config_file`

File: tests/c_languages/AStyleLintBearTest.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+from bears.c_languages.AStyleLintBear import AStyleLintBear
+from coalib.testing.LocalBearTestHelper import verify_local_bear
+
+test_file1 = """
+int main()
+{
+    int a = 19;
+    return 0;
+}"""
```
Comment (jayvdb) [inline tests/c_languages/AStyleLintBearTest.py]: please add a line feed before `"""` so the outputted file ends with a line feed.
(Otherwise the linter may create an error about no EOL and EOF.

File: .travis.yml (hunk 13 above)
Comment (ivange94) [inline .travis.yml]: I just googled how to install dependencies on travis and I got that. But still didn't work. My builds are failing.

Comment (ivange94) [inline .travis.yml]: It was suggested that my builds are failing on travis because it does not have astyle installed

File: bears/c_languages/AStyleLintBear.py (hunk 14 above)
Comment (ivange94) [inline bears/c_languages/AStyleLintBear.py]: I think its just the bracket style. At least thats what the documentation says. But astyle itself just call it `style` and the command line parameter is `style`

File: bears/c_languages/AStyleLintBear.py (hunk 15 above)
Comment (ivange94) [inline bears/c_languages/AStyleLintBear.py]: a lot of of bears still specify the the linter specific config file. It gets me confused. For example https://github.com/coala/coala-bears/blob/master/bears/python/PyLintBear.py#L38

Not sure though but I think the coala config file is called using `--config-file` or `--config` which is different from the option astyle uses to call it's configuration file. So I don't understand how the two relate.

Comment (Makman2) [inline bears/c_languages/AStyleLintBear.py]: nope you define the flag in `create_arguments` :)

The `config_file` parameter contains the full path to the generated config file in `generate_config`.

Anyway coala still supports setting configuration files specific to that tool, so people can migrate easier. You are right, ideally we don't want to allow those, but often those contain way more options than coala currently supports.

This changes with "aspects", where we want to streamline settings and options, and there we won't allow passing some `astyle_config` files (or config files used by other tools).

**Comment by Makman2** [comment]
@ivange94 wanna continue? I think there's not much left for merge. I hope we don't have over-reviewed you, we just try to make the best bears we can :D

**Comment by ivange94** [comment]
@Makman2 sorry about the delay. Will look at this in the morning. Its already 11 pm here and I've been working all day. Been really busy with other stuffs and still haven't figured out why travis and cirlce ci are failing on github. It doesn't locally.

**Comment by Makman2** [comment]
@ivange94 ah sure take your time, just wanted to ask :)

File: bears/c_languages/AStyleLintBear.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    LANGUAGES = {'C', 'C++', 'Objective-C', 'C#', 'Java'}
+    REQUIREMENTS = {DistributionRequirement(apt_get='astyle')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    CAN_FIX = {'Formatting'}
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         astyle_bracket_style: str='',
```
Comment (jayvdb) [inline bears/c_languages/AStyleLintBear.py]: The default here should be 'allman', and then it is self-documenting, and the note below isnt necessary.
Also that means that if astyle releases a version with the default changed, coala users are not immediately impacted.

File: tests/c_languages/AStyleLintBearTest.py (hunk 16 above)
Comment (jayvdb) [inline tests/c_languages/AStyleLintBearTest.py]: I see you have added a `\n` before the `"""`.  What I meant, and sorry for not being more clear, was to use
```
"""
...
}
"""
```

File: tests/c_languages/AStyleLintBearTest.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+from bears.c_languages.AStyleLintBear import AStyleLintBear
+from coalib.testing.LocalBearTestHelper import verify_local_bear
+
+test_file1 = """
+int main()
+{
+    int a = 19;
+    return 0;
+}\n"""
```
Comment (jayvdb) [inline tests/c_languages/AStyleLintBearTest.py]: Please use
use
```
"""
...
}
"""
```

File: tests/c_languages/AStyleLintBearTest.py [hunk 19]
```diff
@@ -0,0 +1,72 @@
+import os
+
+from bears.c_languages.AStyleLintBear import AStyleLintBear
+from coalib.testing.LocalBearTestHelper import verify_local_bear
+
+test_file1 = """
```
Comment (jayvdb) [inline tests/c_languages/AStyleLintBearTest.py]: could you create more descriptive variable names; instead of `test_file1`, maybe `test_good_file_with_assignment`

File: tests/c_languages/AStyleLintBearTest.py [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+                    invalid_files=(test_file4,),
+                    settings={'use_spaces': True,
+                              'indent_size': 4}
+                )
+
+AStyleLintBearTestUseOptionsFile = verify_local_bear(
+                    AStyleLintBear,
+                    valid_files=(test_file1,),
+                    invalid_files=(test_file2,),
+                    settings={'astyle_config': get_testfile_path(
```
Comment (jayvdb) [inline tests/c_languages/AStyleLintBearTest.py]: optional, but preferred, this config file could be generated at runtime.

File: bears/c_languages/AStyleLintBear.py (hunk 17 above)
Comment (ivange94) [inline bears/c_languages/AStyleLintBear.py]: I don't think that is right. If I set the default to allman it means the linter will always be run with command line option --style=allman. Users may not want that. Command line options take precedence over config file options. So If a user had something like --stye=java inside his config file, it will not work because the linter is always run with --style=allman which takes precedence over the config file.

Comment (ivange94) [inline bears/c_languages/AStyleLintBear.py]: I am very certain of this because when I was implementing this long time ago, I set the default to allman and my tests that used config files kept failing. Then I realized that this happened because the config file options were been ignored since they were specified from the command line.

File: tests/c_languages/AStyleLintBearTest.py (hunk 20 above)
Comment (ivange94) [inline tests/c_languages/AStyleLintBearTest.py]: Can you please direct me to a linter that does it the way you suggest? If you noticed in my bear I removed the astyle_config as you commented. I was following the linter tutorial and they had that option but you said its not necessary. But I looked at a lot of linters and they still do that. So can you please direct me to a linter does it the way you want?

File: tests/c_languages/AStyleLintBearTest.py (hunk 18 above)
Comment (ivange94) [inline tests/c_languages/AStyleLintBearTest.py]: ok

File: tests/c_languages/AStyleLintBearTest.py (hunk 19 above)
Comment (ivange94) [inline tests/c_languages/AStyleLintBearTest.py]: ok

File: tests/c_languages/AStyleLintBearTest.py (hunk 20 above)
Comment (jayvdb) [inline tests/c_languages/AStyleLintBearTest.py]: Sorry ... I mean that instead of adding "tests/c_languages/astyle_test_files/astylerc_file" to the repo, this could be a generated temporary file. `CheckstyleBearTest.known_path_test` has an example of using `NamedTemporaryFile` to avoid a separate file for the config.

**Comment by ivange94** [comment]
I'm not working on this again. Can't find the time to complete it.
